### PR TITLE
feat(chisel): improve inspector

### LINKED ISSUE #4130: Chisel - type(int[N]).min is incorrect
### Component

Chisel

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

chisel 0.1.0

### What command(s) is the bug in?

chisel

### Operating System

None

### Describe the bug

See the following screenshot for some examples.
image

It appears that calculating the minimum value for signed integer types is incorrect.

### PR DESCRIPTION
**Pull request by DaniPopes**
## Motivation

Currently many expressions are missing from the `Type` enum which handles expressions to be inspected
Fixes #4130

TLDR:
- Can inspect all Solidity expressions, excluding tuples, with the correct type. This includes, but not limited to:
  - All [global variables](https://docs.soliditylang.org/en/latest/cheatsheet.html#global-variables), with correct abi.decode, type(T).max etc types
  - All type associated methods (array.length, function.address)
  - All assignments, which also get added to the final code
- Inspected arithmetic operations will return int if at least 1 of the operands is also an int

## Solution

- Add missing expressions
- Properly handle functions, accesses and tuples
- Refactor global variable logic to another method
- Clean up 73 levels of indentation functions

Listed in the code

- ~~Continue inspecting if `inspectoor` variable fails~~
  - ~~Any tuple (or empty return) fails compilation due to it not being able to be encoded in `inspectoor`~~
  - This is getting already too big, I think this is better for another PR
- [x] ~~Encode array length and relative offset for dynamic arrays~~ Array literals as `Type::FixedArray`
- [x] If assignment expressions return Some and get "inspected", the assignment does not happen
- [x] Exhaustively match on `from_expression` (blocked by the above)
- [x] Use `Int(n)` for arithmetic operations

Might not be needed or feasible:
- (?) Rational numbers (no corresponding ParamType ?)

### CODE REVIEW

File: chisel/src/executor.rs [hunk 1]
```diff
@@ -376,82 +395,129 @@ impl Type {
...
-                        let num = if num.is_empty() { 0usize } else { num.parse().ok()? };
-                        let exp = if exp.is_empty() { 0u32 } else { exp.parse().ok()? };
-                        Some(num * 10usize.pow(exp))
+                Self::from_expression(expr).and_then(|ty| {
+                    let ty = Box::new(ty);
+                    let num = num.as_deref().and_then(parse_number_literal);
+                    if let Some(num) = num {
+                        // overflow check
+                        const MAX_U64_AS_U256: U256 = U256([0, 0, 0, usize::MAX as u64]);
+                        if num > MAX_U64_AS_U256 {
```
Comment (mattsse) [inline chisel/src/executor.rs:412]: last time we changed U256 constants in ethers we got burned, so I'd appreciate a simple sanity check for this constant.

File: cli/tests/it/cmd.rs [hunk 2]
```diff
@@ -368,7 +368,7 @@ forgetest_init!(can_clean_config, |prj: TestProject, mut cmd: TestCommand| {
     cmd.assert_non_empty_stdout();
```
Comment (mattsse) [inline cli/tests/it/cmd.rs:368]: please submit the clippy/style changes for `cli/tests` separately

**Review by mattsse — CHANGES_REQUESTED** [review]
code-wise this lgtm, 
not too familiar with the issue, can we get a test case for this perhaps?

wdyt @clabby @rkrasiuk

**Comment by DaniPopes** [comment]
Not sure how I am supposed to configure Solc to run tests ..

File: cli/tests/it/cmd.rs [hunk 3]
```diff
@@ -368,7 +368,7 @@ forgetest_init!(can_clean_config, |prj: TestProject, mut cmd: TestCommand| {
...
```
Comment (mattsse) [inline cli/tests/it/cmd.rs:369]: didn't we just merge a PR that changed the format! s?

In any case, please undo because unrelated here.

**Review by mattsse — COMMENTED** [review]
a lot of unrelated changes which makes this large PR harder to review, please undo unrelated changes and submit separately if applicable.

File: cli/tests/it/cmd.rs (hunk 3 above)
Comment (DaniPopes) [reply cli/tests/it/cmd.rs:369]: this was reverted like 10 commits ago , this isnt the current diff

**Comment by clabby** [comment]
> code-wise this lgtm, not too familiar with the issue, can we get a test case for this perhaps?
> 
> wdyt @clabby @rkrasiuk

Busy with some work stuff, but will review this & #4147 later on tonight or tomorrow morning! Appreciate the improvements @DaniPopes 🚀

**Comment by rkrasiuk** [comment]
same as what @mattsse said, the code looks good. deferring to @clabby for approval

File: chisel/src/executor.rs [hunk 4]
```diff
@@ -458,15 +458,15 @@ impl Type {
             // pt::Expression::PostDecrement(_, inner) |       // <inner>--
             // pt::Expression::PreIncrement(_, inner) |        // ++<inner>
             // pt::Expression::PostIncrement(_, inner) |       // <inner>++
-            // pt::Expression::Assign(_, inner, _) |           // <inner> = ...
```
Comment (x676f64) [inline chisel/src/executor.rs]: Should remove this change from your commit.

**Comment by x676f64** [comment]
> code-wise this lgtm, not too familiar with the issue, can we get a test case for this perhaps?
> 
> wdyt @clabby @rkrasiuk

The initial impetus for this PR was https://github.com/foundry-rs/foundry/issues/4130 which includes some failing test cases pertaining to signed integers.

**Review by clabby — COMMENTED** [review]
This is looking pretty good code-wise, but experiencing some issues with inspecting on commit `76acc498ea29cbddf5b649467ff25131b0ffcf2b`. On my machine, it crashes every time a variable or expression is inspected (regardless of the expression type)

Env:
- **OS:** Mac OS
- **Arch**: ARM (Apple Silicon)
Screenshot 2023-01-24 at 2 40 14 PM

**Comment by mattsse** [comment]
we merged two other chisel PRs yesterday,

just to rule it out, this does __not__ panic with nightly chisel?

**Comment by DaniPopes** [comment]
> On my machine, it crashes every time a variable or expression is inspected (regardless of the expression type)

I introduced infinite recursion in 799e2b42653fa88b27d9bb92e404b430dfc3611f by mistake because I didn't merge the `try_as_ethabi` match arms properly. Fixed in ed8683863ba6d5a24659241bd8cc0699bdf48367, sorry about that

**Review by clabby — APPROVED** [review]
This looks good to me, thanks for the improvements! There are some commented out blocks (specifically in the `executor` mod) that could use some cleaning up, but that's just a stylistic nit.

File: chisel/src/executor.rs [hunk 5]
```diff
@@ -657,100 +888,288 @@ impl Type {
         }
     }
 
+    /// Infers the type from a variable's type
+    fn infer_var_expr(
+        expr: &pt::Expression,
+        intermediate: Option<&IntermediateOutput>,
+        custom_type: &mut Vec<String>,
+        // contract_name: Option<String>,
+        // intermediate_contract: Option<&IntermediateContract>,
```
Comment (mattsse) [inline chisel/src/executor.rs:897]: no longer used? can remove?

File: chisel/src/executor.rs [hunk 6]
```diff
@@ -657,100 +888,288 @@ impl Type {
...
+                let name = &ident.name;
+
+                if let Some(intermediate) = intermediate {
+                    // if let Some(intermediate_contract) = intermediate_contract {
+                    //     if intermediate_contract.struct_definitions.contains_key(name) {
+                    //         custom_type.clear();
+                    //         custom_type.push(name.clone());
+                    //         return Self::infer_custom_type(intermediate, custom_type,
+                    // contract_name)     }
+                    // }
```
Comment (mattsse) [inline chisel/src/executor.rs:912]: can be removed?

**Review by mattsse — APPROVED** [review]
pending, code comment nits otherwise lgtm

tysm

**Review by mattsse — APPROVED** [review]
gg champ
